### ISSUE TITLE
Custom test main file with allModules

### ISSUE THREAD

**Issue by shishkander** [dlang/dub#661]
**tl;dr** dub does a great job figuring out all modules [here](https://github.com/D-Programming-Language/dub/blob/c7fded73f74f596448a094d773053800dd703f74/source/dub/dub.d#L401). Can it be exposed to custom test main file?

Motivation:

The iteration over all modules is really awesome, and after #424 can be done, it'd be perfect. So is [tested](https://github.com/rejectedsoftware/tested) package. I want to use a custom `TestResultWriter`. Specifically, I wanted to add printing of nice stacktraces in case of error using backtrace package.

**Current solution:**

I'm relying on current implementation of dub which names the generated file `dub_test_root.d` which I then import from my custom test_main_file like this:

```
module test;
class CustomTestResultWriter{...}
version (unittest) {

    shared static this() {
        import core.runtime;

        Runtime.moduleUnitTester = () => true;
    }

    void main() {
        import std.exception;
        import tested;
        import dub_test_root;

        enforce(runUnitTests!(dub_test_root.allModules)(new CustomTestResultWriter()));
    }
}
```

Any better approaches?

**Comment by WebFreak001** [comment]
You can hook into the test runner using just a module constructor from a dependency, like what [silly](https://code.dlang.org/packages/silly) does:

```d
shared static this() {
	import core.runtime : Runtime, UnitTestResult;

	Runtime.extendedModuleUnitTester = function () {
		int passed, failed;
		// run tests here

		return UnitTestResult(passed + failed, passed, false, false);
	}
}
```

as targetType: sourceLibrary
